File: src/larkdesk/listen.py
"""Listen for inbound text pushes on the desktop-session frontier socket."""

from __future__ import annotations

import json
import shlex
import subprocess
import sys
from collections import deque
from dataclasses import dataclass
from typing import Callable

from larkdesk.contacts import Contact, list_contacts
from larkdesk.frontier import decode_frontier_packet, iter_frontier_frames
from larkdesk.gateway import decode_fields
from larkdesk.search import SearchError
from larkdesk.send import MSG_TYPE_TEXT, SendError, resolve_person
from larkdesk.session import AuthMaterial, load_session

CMD_PUSH_MESSAGES = 6
CMD_PUSH_MESSAGES_V2 = 5065
_SEEN_MAX = 512


class ListenError(RuntimeError):
    """Inbound listen could not start or decode."""


@dataclass(frozen=True)
class InboundEvent:
    ts: int
    chat_id: str
    message_id: str
    from_id: str
    from_name: str
    text: str
    self: bool
    cmd: int = CMD_PUSH_MESSAGES

    def as_dict(self) -> dict[str, object]:
        frm: dict[str, str] = {"id": self.from_id}
        if self.from_name:
            frm["name"] = self.from_name
        return {
            "ts": self.ts,
            "chat_id": self.chat_id,
            "message_id": self.message_id,
            "from": frm,
            "text": self.text,
            "self": self.self,
        }


def _utf8(data: bytes) -> str | None:
    try:
        text = data.decode("utf-8")
    except UnicodeDecodeError:
        return None
    return text if text else None


def _printable(text: str) -> bool:
    return bool(text) and all(ch.isprintable() or ch in "\n\r\t" for ch in text)


def extract_plain_text(buf: bytes) -> str:
    """Pull RichText.innerText, else the longest non-id string in the tree."""
    inner: list[str] = []
    strings: list[str] = []

    def walk(raw: bytes) -> None:
        fields = decode_fields(raw)
        tags = {f for f, _k, _v in fields}
        for field, kind, value in fields:
            if kind != "bytes" or not isinstance(value, (bytes, bytearray)):
                continue
            chunk = bytes(value)
            text = _utf8(chunk)
            if text and _printable(text):
                strings.append(text)
                if field == 2 and 3 in tags and text.strip() and not text.isdigit():
                    inner.append(text)
            if chunk:
                walk(chunk)

    walk(buf)
    for text in inner:
        if text.strip():
            return text
    cands = [s for s in strings if not s.isdigit() and _printable(s) and len(s) >= 1]
    return max(cands, key=len) if cands else ""


def _message_from_fields(buf: bytes) -> dict[str, object] | None:
    fields = decode_fields(buf)
    if not fields:
        return None
    by: dict[int, list] = {}
    for field, kind, value in fields:
        by.setdefault(field, []).append((kind, value))
    type_hit = by.get(2, [(None, None)])[0]
    if type_hit[0] != "varint":
        return None
    msg_type = int(type_hit[1])  # type: ignore[arg-type]
    mid = ""
    from_id = ""
    chat_id = ""
    ts = 0
    content = b""
    if 1 in by and by[1][0][0] == "bytes":
        mid = _utf8(bytes(by[1][0][1])) or ""  # type: ignore[arg-type]
    if 3 in by and by[3][0][0] == "bytes":
        from_id = _utf8(bytes(by[3][0][1])) or ""  # type: ignore[arg-type]
    if 4 in by and by[4][0][0] == "varint":
        ts = int(by[4][0][1])  # type: ignore[arg-type]
    if 5 in by and by[5][0][0] == "bytes":
        content = bytes(by[5][0][1])  # type: ignore[arg-type]
    if 10 in by and by[10][0][0] == "bytes":
        chat_id = _utf8(bytes(by[10][0][1])) or ""  # type: ignore[arg-type]
    return {
        "id": mid,
        "type": msg_type,
        "from_id": from_id,
        "ts": ts,
        "content": content,
        "chat_id": chat_id,
    }


def parse_push_messages(payload: bytes) -> list[dict[str, object]]:
    """messages.PushMessagesRequest: repeated { id=1, message=2 }."""
    out: list[dict[str, object]] = []
    for field, kind, value in decode_fields(payload):
        if kind != "bytes" or not isinstance(value, (bytes, bytearray)):
            continue
        item = bytes(value)
        direct = _message_from_fields(item)
        if direct and direct.get("type") == MSG_TYPE_TEXT:
            out.append(direct)
            continue
        inner_msg = b""
        inner_id = ""
        for inf, ink, inv in decode_fields(item):
            if inf == 1 and ink == "bytes" and isinstance(inv, (bytes, bytearray)):
                inner_id = _utf8(bytes(inv)) or inner_id
            elif inf == 2 and ink == "bytes" and isinstance(inv, (bytes, bytearray)):
                inner_msg = bytes(inv)
        parsed = _message_from_fields(inner_msg) if inner_msg else None
        if not parsed:
            continue
        if not parsed.get("id"):
            parsed["id"] = inner_id
        if parsed.get("type") == MSG_TYPE_TEXT:
            out.append(parsed)
    return out


def events_from_frame(
    raw: bytes,
    *,
    self_id: str = "",
    names: dict[str, str] | None = None,
) -> list[InboundEvent]:
    pkt = decode_frontier_packet(raw)
    cmd = pkt.get("cmd")
    if cmd not in (CMD_PUSH_MESSAGES, CMD_PUSH_MESSAGES_V2):
        return []
    payload = pkt.get("payload")
    if not isinstance(payload, (bytes, bytearray)):
        return []
    names = names or {}
    events: list[InboundEvent] = []
    for msg in parse_push_messages(bytes(payload)):
        text = extract_plain_text(bytes(msg.get("content") or b""))
        if not text.strip():
            continue
        from_id = str(msg.get("from_id") or "")
        mid = str(msg.get("id") or "")
        events.append(
            InboundEvent(
                ts=int(msg.get("ts") or 0),
                chat_id=str(msg.get("chat_id") or ""),
                message_id=mid,
                from_id=from_id,
                from_name=names.get(from_id, ""),
                text=text,
                self=bool(self_id) and from_id == self_id,
                cmd=int(cmd),
            )
        )
    return events


def _name_map(auth: AuthMaterial) -> dict[str, str]:
    names = {auth.identity.user_id: auth.identity.name}
    try:
        _auth, contacts, _src = list_contacts(auth=auth)
    except Exception:
        return names
    for contact in contacts:
        if contact.id and contact.name:
            names.setdefault(contact.id, contact.name)
    return names


def resolve_listen_filter(who: str, *, auth: AuthMaterial) -> Contact | None:
    who = who.strip()
    if not who:
        return None
    try:
        return resolve_person(who, auth=auth)
    except (SendError, SearchError) as exc:
        raise ListenError(str(exc)) from exc


def _run_hook(command: str, payload: dict[str, object]) -> None:
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    try:
        subprocess.run(shlex.split(command), input=data, check=False)
    except Exception as exc:
        print(f"larkdesk: exec {exc}", file=sys.stderr)


async def listen_loop(
    auth: AuthMaterial,
    *,
    allow_from_id: str = "",
    on_event: Callable[[InboundEvent], None],
    names: dict[str, str] | None = None,
) -> None:
    seen: deque[str] = deque(maxlen=_SEEN_MAX)
    seen_set: set[str] = set()
    async for raw in iter_frontier_frames(auth):
        for event in events_from_frame(raw, self_id=auth.identity.user_id, names=names):
            if allow_from_id and event.from_id != allow_from_id:
                continue
            if event.message_id:
                if event.message_id in seen_set:
                    continue
                if len(seen) == seen.maxlen:
                    old = seen[0]
                    seen_set.discard(old)
                seen.append(event.message_id)
                seen_set.add(event.message_id)
            on_event(event)


def run_listen(
    *,
    from_who: str = "",
    exec_cmd: str = "",
    auth: AuthMaterial | None = None,
) -> None:
    session = auth or load_session()
    if not session.cookies:
        raise ListenError("no decrypted session cookie")
    allow: Contact | None = None
    if from_who.strip():
        allow = resolve_listen_filter(from_who, auth=session)
    names = _name_map(session)

    def emit(event: InboundEvent) -> None:
        payload = event.as_dict()
        sys.stdout.write(json.dumps(payload, ensure_ascii=False) + "\n")
        sys.stdout.flush()
        if exec_cmd:
            _run_hook(exec_cmd, payload)

    print(
        f"larkdesk: listening as {session.identity.name}",
        file=sys.stderr,
    )
    import asyncio

    asyncio.run(
        listen_loop(
            session,
            allow_from_id=allow.id if allow else "",
            on_event=emit,
            names=names,
        )
    )

File: internal/listen/listen.go
package listen

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
	"unicode"

	"github.com/TensorFu/larkdesk/internal/contacts"
	"github.com/TensorFu/larkdesk/internal/frontier"
	"github.com/TensorFu/larkdesk/internal/pb"
	"github.com/TensorFu/larkdesk/internal/send"
	"github.com/TensorFu/larkdesk/internal/session"
)

const (
	CmdPushMessages   = 6
	CmdPushMessagesV2 = 5065
	seenMax           = 512
)

type Event struct {
	TS        uint64 `json:"ts"`
	ChatID    string `json:"chat_id"`
	MessageID string `json:"message_id"`
	FromID    string `json:"-"`
	FromName  string `json:"-"`
	Text      string `json:"text"`
	Self      bool   `json:"self"`
	Cmd       uint64 `json:"-"`
}

func (e Event) MarshalJSON() ([]byte, error) {
	frm := map[string]string{"id": e.FromID}
	if e.FromName != "" {
		frm["name"] = e.FromName
	}
	return json.Marshal(map[string]any{
		"ts":         e.TS,
		"chat_id":    e.ChatID,
		"message_id": e.MessageID,
		"from":       frm,
		"text":       e.Text,
		"self":       e.Self,
	})
}

func printable(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r == '\n' || r == '\r' || r == '\t' {
			continue
		}
		if !unicode.IsPrint(r) {
			return false
		}
	}
	return true
}

func ExtractPlainText(buf []byte) string {
	var inner, stringsFound []string
	var walk func([]byte)
	walk = func(raw []byte) {
		fields := pb.DecodeFields(raw)
		tags := map[int]bool{}
		for _, f := range fields {
			tags[f.Num] = true
		}
		for _, f := range fields {
			if f.Wire != "bytes" {
				continue
			}
			if s, ok := utf8str(f.Bytes); ok && printable(s) {
				stringsFound = append(stringsFound, s)
				if f.Num == 2 && tags[3] && strings.TrimSpace(s) != "" && !isDigits(s) {
					inner = append(inner, s)
				}
			}
			if len(f.Bytes) > 0 {
				walk(f.Bytes)
			}
		}
	}
	walk(buf)
	for _, t := range inner {
		if strings.TrimSpace(t) != "" {
			return t
		}
	}
	best := ""
	for _, s := range stringsFound {
		if !isDigits(s) && printable(s) && len(s) > len(best) {
			best = s
		}
	}
	return best
}

func utf8str(b []byte) (string, bool) {
	s := string(b)
	if s == "" {
		return "", false
	}
	return s, true
}

func isDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if !unicode.IsDigit(r) {
			return false
		}
	}
	return true
}

type msg struct {
	ID      string
	Type    uint64
	FromID  string
	TS      uint64
	Content []byte
	ChatID  string
}

func messageFromFields(buf []byte) (msg, bool) {
	fields := pb.DecodeFields(buf)
	if len(fields) == 0 {
		return msg{}, false
	}
	by := map[int][]pb.Field{}
	for _, f := range fields {
		by[f.Num] = append(by[f.Num], f)
	}
	t, ok := by[2]
	if !ok || t[0].Wire != "varint" {
		return msg{}, false
	}
	m := msg{Type: t[0].Int}
	if f, ok := by[1]; ok && f[0].Wire == "bytes" {
		if s, good := pb.UTF8Printable(f[0].Bytes); good {
			m.ID = s
		} else {
			m.ID = string(f[0].Bytes)
		}
	}
	if f, ok := by[3]; ok && f[0].Wire == "bytes" {
		if s, good := pb.UTF8Printable(f[0].Bytes); good {
			m.FromID = s
		}
	}
	if f, ok := by[4]; ok && f[0].Wire == "varint" {
		m.TS = f[0].Int
	}
	if f, ok := by[5]; ok && f[0].Wire == "bytes" {
		m.Content = f[0].Bytes
	}
	if f, ok := by[10]; ok && f[0].Wire == "bytes" {
		if s, good := pb.UTF8Printable(f[0].Bytes); good {
			m.ChatID = s
		}
	}
	return m, true
}

func ParsePushMessages(payload []byte) []msg {
	var out []msg
	for _, f := range pb.DecodeFields(payload) {
		if f.Wire != "bytes" {
			continue
		}
		if direct, ok := messageFromFields(f.Bytes); ok && direct.Type == send.MsgTypeText {
			out = append(out, direct)
			continue
		}
		innerID := ""
		var innerMsg []byte
		for _, inf := range pb.DecodeFields(f.Bytes) {
			if inf.Num == 1 && inf.Wire == "bytes" {
				if s, ok := pb.UTF8Printable(inf.Bytes); ok {
					innerID = s
				}
			} else if inf.Num == 2 && inf.Wire == "bytes" {
				innerMsg = inf.Bytes
			}
		}
		if len(innerMsg) == 0 {
			continue
		}
		parsed, ok := messageFromFields(innerMsg)
		if !ok {
			continue
		}
		if parsed.ID == "" {
			parsed.ID = innerID
		}
		if parsed.Type == send.MsgTypeText {
			out = append(out, parsed)
		}
	}
	return out
}

func EventsFromFrame(raw []byte, selfID string, names map[string]string) []Event {
	fr := frontier.DecodeFrame(raw)
	if fr.Cmd != CmdPushMessages && fr.Cmd != CmdPushMessagesV2 {
		return nil
	}
	if len(fr.Payload) == 0 {
		return nil
	}
	var events []Event
	for _, m := range ParsePushMessages(fr.Payload) {
		text := ExtractPlainText(m.Content)
		if strings.TrimSpace(text) == "" {
			continue
		}
		events = append(events, Event{
			TS:        m.TS,
			ChatID:    m.ChatID,
			MessageID: m.ID,
			FromID:    m.FromID,
			FromName:  names[m.FromID],
			Text:      text,
			Self:      selfID != "" && m.FromID == selfID,
			Cmd:       fr.Cmd,
		})
	}
	return events
}

func nameMap(auth session.Auth) map[string]string {
	names := map[string]string{auth.Identity.UserID: auth.Identity.Name}
	cs, _, err := contacts.List(auth, "")
	if err != nil {
		return names
	}
	for _, c := range cs {
		if c.ID != "" && c.Name != "" {
			if _, ok := names[c.ID]; !ok {
				names[c.ID] = c.Name
			}
		}
	}
	return names
}

func Run(auth session.Auth, fromWho, execCmd string) error {
	if !auth.HasSessionCookie() {
		return fmt.Errorf("no decrypted session cookie")
	}
	allowID := ""
	if strings.TrimSpace(fromWho) != "" {
		p, err := send.ResolvePerson(auth, fromWho)
		if err != nil {
			return err
		}
		allowID = p.ID
	}
	names := nameMap(auth)
	fmt.Fprintf(os.Stderr, "larkdesk: listening as %s\n", auth.Identity.Name)

	seen := []string{}
	seenSet := map[string]bool{}
	enc := json.NewEncoder(os.Stdout)
	enc.SetEscapeHTML(false)

	for {
		conn, err := frontier.Dial(auth)
		if err != nil {
			time.Sleep(3 * time.Second)
			continue
		}
		for {
			_, raw, err := conn.ReadMessage()
			if err != nil {
				conn.Close()
				time.Sleep(3 * time.Second)
				break
			}
			raw = frontier.MaybeGunzip(raw)
			for _, ev := range EventsFromFrame(raw, auth.Identity.UserID, names) {
				if allowID != "" && ev.FromID != allowID {
					continue
				}
				if ev.MessageID != "" {
					if seenSet[ev.MessageID] {
						continue
					}
					if len(seen) == seenMax {
						delete(seenSet, seen[0])
						seen = seen[1:]
					}
					seen = append(seen, ev.MessageID)
					seenSet[ev.MessageID] = true
				}
				if err := enc.Encode(ev); err != nil {
					return err
				}
				if execCmd != "" {
					b, _ := json.Marshal(ev)
					cmd := exec.Command("sh", "-c", execCmd)
					cmd.Stdin = strings.NewReader(string(b))
					cmd.Stdout = os.Stderr
					cmd.Stderr = os.Stderr
					_ = cmd.Run()
				}
			}
		}
	}
}
